File: church/church/report/tithe_report/tithe_report.js
// Copyright (c) 2026, Your Organization
// For license information, please see license.txt

frappe.query_reports["Tithe Report"] = {
    "filters": [
        {
            "fieldname": "from_date",
            "label": __("🌸 From Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.add_months(frappe.datetime.get_today(), -1),
            "reqd": 1
        },
        {
            "fieldname": "to_date",
            "label": __("🌼 To Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.get_today(),
            "reqd": 1
        },
        {
            "fieldname": "batch_id",
            "label": __("📦 Batch"),
            "fieldtype": "Link",
            "options": "Tithe Batch Update",
            "get_query": function() {
                return {
                    filters: {
                        "docstatus": 1
                    }
                };
            }
        },
        {
            "fieldname": "member_id",
            "label": __("👤 Member"),
            "fieldtype": "Link",
            "options": "Member",
            "get_query": function() {
                return {
                    filters: {
                        "member_status": ["in", ["Active", "Inactive", "Left", "Transferred"]]
                    }
                };
            }
        },
        {
            "fieldname": "member_status",
            "label": __("⭐ Member Status"),
            "fieldtype": "Select",
            "options": ["All", "Active", "Inactive", "Left", "Transferred"],
            "default": "All"
        },
        {
            "fieldname": "type",
            "label": __("🌟 Source"),
            "fieldtype": "Select",
            "options": ["", "Worker", "Member"],
            "default": ""
        },
        {
            "fieldname": "branch",
            "label": __("🏢 Branch"),
            "fieldtype": "Link",
            "options": "Branch"
        },
        {
            "fieldname": "currency",
            "label": __("💵 Currency"),
            "fieldtype": "Link",
            "options": "Currency"
        },
        {
            "fieldname": "base_currency",
            "label": __("🏦 Base Currency"),
            "fieldtype": "Link",
            "options": "Currency",
            "default": "USD"
        }
    ],
    
    // Formatter with styling
    "formatter": function(value, row, column, data, default_formatter) {
        value = default_formatter(value, row, column, data);
        
        const colors = {
            active: "#10B981",
            inactive: "#6B7280",
            left: "#EF4444",
            transferred: "#F59E0B",
            subtotal: "#FEF3C7",
            subtotal_text: "#92400E",
            grand_total: "#D1FAE5",
            grand_total_text: "#065F46"
        };
        
        // Special styling for subtotal rows
        if (row && row.is_subtotal) {
            if (column.fieldname == "type") {
                return `<div style="background-color: ${colors.subtotal}; color: ${colors.subtotal_text}; padding: 4px 8px; border-radius: 6px; font-weight: bold; text-align: center;">✨ SUBTOTAL ✨</div>`;
            }
            if (column.fieldname == "receipt_reference") {
                return `<span style="color: ${colors.subtotal_text}; font-style: italic;">${value}</span>`;
            }
            if (column.fieldname == "full_name") {
                return `<strong style="color: ${colors.subtotal_text};">${value}</strong>`;
            }
            if (["amount_paid", "worker_tithe", "member_tithe", "total_base_currency"].includes(column.fieldname)) {
                return `<strong style="color: ${colors.subtotal_text};">${value}</strong>`;
            }
        }
        
        // Special styling for grand total row
        if (row && row.is_grand_total) {
            if (column.fieldname == "member_id") {
                return `<div style="background: linear-gradient(135deg, ${colors.grand_total}, #A7F3D0); padding: 6px 12px; border-radius: 8px; font-weight: bold; font-size: 1.1em; text-align: center;">🎯 ${value} 🎯</div>`;
            }
            if (column.fieldname == "type") {
                return `<div style="background-color: ${colors.grand_total}; color: ${colors.grand_total_text}; padding: 4px 8px; border-radius: 6px; font-weight: bold; text-align: center;">🏆 GRAND TOTAL 🏆</div>`;
            }
            if (column.fieldname == "receipt_reference") {
                return `<strong style="color: ${colors.grand_total_text};">${value}</strong>`;
            }
            if (["amount_paid", "worker_tithe", "member_tithe", "total_base_currency"].includes(column.fieldname)) {
                return `<strong style="color: ${colors.grand_total_text}; font-size: 1.05em;">${value}</strong>`;
            }
        }
        
        // Add emojis and styling to regular rows
        if (column.fieldname == "type" && value && !row.is_subtotal && !row.is_grand_total) {
            if (value == "Worker") value = "🙏 " + value;
            if (value == "Member") value = "🏡 " + value;
        }
        
        if (column.fieldname == "currency" && value && !row.is_subtotal && !row.is_grand_total) {
            value = "💵 " + value;
        }
        
        if (column.fieldname == "receipt_reference" && value && !row.is_subtotal && !row.is_grand_total) {
            value = "🧾 " + value;
        }
        
        if (column.fieldname == "member_status" && value) {
            const statusColors = {
                "Active": colors.active,
                "Inactive": colors.inactive,
                "Left": colors.left,
                "Transferred": colors.transferred
            };
            const statusEmojis = {
                "Active": "🟢",
                "Inactive": "⚪",
                "Left": "🔴",
                "Transferred": "🔄"
            };
            const color = statusColors[value] || colors.inactive;
            const emoji = statusEmojis[value] || "⭐";
            value = `<span style="color: ${color}; font-weight: 500;">${emoji} ${value}</span>`;
        }
        
        if (column.fieldname == "batch_name" && value) {
            value = `<span style="font-family: monospace; font-size: 0.9em;">📦 ${value}</span>`;
        }
        
        return value;
    },
    
    "tree": false,
    
    "onload": function(report) {
        frappe.show_alert({
            message: __("🎶 Counting blessings and tithes from submitted batches... ✨"),
            indicator: 'green'
        }, 3);
        
        // Add custom Excel button
        this.add_custom_excel_button(report);
        
        // Add friendly greeting
        setTimeout(() => {
            this.add_friendly_greeting();
        }, 500);
    },
    
    "refresh": function(report) {
        // Re-add button on refresh
        this.add_custom_excel_button(report);
    },
    
    "add_custom_excel_button": function(report) {
        // Wait for the page to be ready
        setTimeout(() => {
            // Check if button already exists
            if ($(".custom-excel-btn-styled").length > 0) {
                return;
            }
            
            // Find the report toolbar
            let toolbar = $(".page-actions .btn-group");
            
            if (toolbar.length === 0) {
                toolbar = $(".page-actions");
            }
            
            if (toolbar.length > 0) {
                // Create styled Excel button
                const excelButton = $(`
                    <button class="btn btn-success custom-excel-btn-styled" style="
                        background: linear-gradient(135deg, #10B981 0%, #059669 100%);
                        border: none;
                        color: white;
                        font-weight: 500;
                        margin-left: 8px;
                        padding: 6px 16px;
                        border-radius: 6px;
                        display: inline-flex;
                        align-items: center;
                        gap: 6px;
                        box-shadow: 0 1px 2px rgba(0,0,0,0.05);
                        transition: all 0.2s ease;
                    ">
                        📊 Export to Excel
                    </button>
                `);
                
                // Add click handler
                excelButton.on("click", () => {
                    this.export_to_excel(report);
                });
                
                // Add hover effect
                excelButton.hover(
                    function() {
                        $(this).css({
                            'background': 'linear-gradient(135deg, #059669 0%, #047857 100%)',
                            'transform': 'translateY(-1px)',
                            'box-shadow': '0 4px 6px rgba(0,0,0,0.1)'
                        });
                    },
                    function() {
                        $(this).css({
                            'background': 'linear-gradient(135deg, #10B981 0%, #059669 100%)',
                            'transform': 'translateY(0)',
                            'box-shadow': '0 1px 2px rgba(0,0,0,0.05)'
                        });
                    }
                );
                
                // Append to toolbar
                toolbar.append(excelButton);
            }
        }, 100);
    },
    
    "export_to_excel": function(report) {
        // Show loading message
        frappe.show_alert({
            message: __("✨ Preparing your Excel report..."),
            indicator: 'green'
        }, 2);
        
        // Get current filter values
        const filters = report.filters;
        const filter_values = {};
        
        if (filters) {
            for (let key in filters) {
                if (filters.hasOwnProperty(key) && filters[key] && filters[key].get_value) {
                    filter_values[key] = filters[key].get_value();
                }
            }
        }
        
        // Build URL for export
        const params = {
            doctype: "Tithe Report",
            report_name: "Tithe Report",
            format: "Excel",
            _lang: frappe.boot.lang,
            from_date: filter_values.from_date || frappe.datetime.add_months(frappe.datetime.get_today(), -1),
            to_date: filter_values.to_date || frappe.datetime.get_today(),
            batch_id: filter_values.batch_id || "",
            member_id: filter_values.member_id || "",
            member_status: filter_values.member_status || "All",
            type: filter_values.type || "",
            branch: filter_values.branch || "",
            currency: filter_values.currency || "",
            base_currency: filter_values.base_currency || "USD"
        };
        
        // Create query string
        const queryString = Object.keys(params)
            .map(key => `${encodeURIComponent(key)}=${encodeURIComponent(params[key])}`)
            .join('&');
        
        // Trigger download
        const url = `/api/method/frappe.core.report.report_export.export?${queryString}`;
        
        // Create hidden iframe or anchor for download
        const downloadFrame = document.createElement('iframe');
        downloadFrame.style.display = 'none';
        downloadFrame.src = url;
        document.body.appendChild(downloadFrame);
        
        // Remove iframe after download starts
        setTimeout(() => {
            document.body.removeChild(downloadFrame);
            frappe.show_alert({
                message: __("✅ Excel report generated! Check your downloads folder. 📥"),
                indicator: 'green'
            }, 3);
        }, 2000);
    },
    
    "add_friendly_greeting": function() {
        const report_container = $(".report-view");
        if (report_container.length && !$(".report-greeting").length) {
            report_container.prepend(`
                <div class="alert alert-info report-greeting" style="
                    margin: 10px 20px;
                    background: linear-gradient(135deg, #FEF3C7, #FDE68A);
                    border: none;
                    border-radius: 12px;
                    box-shadow: 0 2px 4px rgba(0,0,0,0.05);
                ">
                    <div style="display: flex; align-items: center; justify-content: space-between;">
                        <div>
                            <span style="font-size: 1.1em; font-weight: 500;">✨ God bless your giving! 🙏</span>
                            <span style="margin-left: 12px; font-size: 0.9em; color: #92400E;">📊 Showing only submitted batches</span>
                        </div>
                        <button type="button" class="close" data-dismiss="alert" aria-label="Close" style="font-size: 1.2em;">
                            <span aria-hidden="true">&times;</span>
                        </button>
                    </div>
                </div>
            `);
        }
    }
};

File: church/church/report/tithe_report/tithe_report.py
# Copyright (c) 2026, Your Organization
# For license information, please see license.txt

from __future__ import unicode_literals
import frappe
from frappe import _
from frappe.utils import flt, getdate, formatdate

def execute(filters=None):
    """
    Main function to execute the Tithe Transaction Report.
    Queries through the parent doctype Tithe Batch Update.
    """
    # Define the columns for the report
    columns = get_columns()
    
    # Fetch data based on filters
    data = get_data(filters)
    
    # Add subtotals for each member and grand total
    data_with_totals = add_subtotals_and_grand_totals(data, filters)
    
    # Prepare summary statistics for the report footer
    summary = get_summary(data_with_totals, filters)
    
    # Add a cute, friendly message if no data found
    if not data:
        frappe.msgprint(_("🎉 Hooray! No tithe transactions found for the selected criteria. Keep shining! ✨"))
        return columns, [], None, None, summary
    
    return columns, data_with_totals, None, None, summary

def get_columns():
    """
    Define the columns for the report.
    """
    return [
        {
            "fieldname": "member_id",
            "label": _("👤 Member ID"),
            "fieldtype": "Link",
            "options": "Member",
            "width": 120
        },
        {
            "fieldname": "full_name",
            "label": _("✨ Full Name"),
            "fieldtype": "Data",
            "width": 200
        },
        {
            "fieldname": "member_status",
            "label": _("⭐ Status"),
            "fieldtype": "Data",
            "width": 100
        },
        {
            "fieldname": "batch_name",
            "label": _("📦 Batch"),
            "fieldtype": "Link",
            "options": "Tithe Batch Update",
            "width": 150
        },
        {
            "fieldname": "branch",
            "label": _("🏢 Branch"),
            "fieldtype": "Link",
            "options": "Branch",
            "width": 150
        },
        {
            "fieldname": "type",
            "label": _("🌟 Source"),
            "fieldtype": "Data",
            "width": 100
        },
        {
            "fieldname": "date",
            "label": _("📅 Date"),
            "fieldtype": "Date",
            "width": 120
        },
        {
            "fieldname": "currency",
            "label": _("💵 Currency"),
            "fieldtype": "Link",
            "options": "Currency",
            "width": 100
        },
        {
            "fieldname": "amount_paid",
            "label": _("💰 Amount Paid"),
            "fieldtype": "Currency",
            "options": "currency",
            "width": 130
        },
        {
            "fieldname": "exchange_rate",
            "label": _("🔄 Exchange Rate"),
            "fieldtype": "Float",
            "precision": 6,
            "width": 110
        },
        {
            "fieldname": "worker_tithe",
            "label": _("🙏 Worker Tithe"),
            "fieldtype": "Currency",
            "options": "currency",
            "width": 130
        },
        {
            "fieldname": "member_tithe",
            "label": _("🏡 Member Tithe"),
            "fieldtype": "Currency",
            "options": "currency",
            "width": 130
        },
        {
            "fieldname": "total_base_currency",
            "label": _("🏦 Total (Base)"),
            "fieldtype": "Currency",
            "width": 150
        },
        {
            "fieldname": "receipt_reference",
            "label": _("🧾 Receipt Ref"),
            "fieldtype": "Data",
            "width": 150
        },
        {
            "fieldname": "other_details",
            "label": _("📝 Details"),
            "fieldtype": "Small Text",
            "width": 250
        }
    ]

def get_conditions(filters):
    """
    Build SQL WHERE conditions based on user filters.
    """
    conditions = ""
    
    if filters.get("from_date"):
        conditions += f" AND tbu.reporting_date >= '{filters.get('from_date')}'"
    
    if filters.get("to_date"):
        conditions += f" AND tbu.reporting_date <= '{filters.get('to_date')}'"
    
    if filters.get("batch_id"):
        conditions += f" AND tbu.name = '{filters.get('batch_id')}'"
    
    if filters.get("member_id"):
        conditions += f" AND tte.member_id = '{filters.get('member_id')}'"
    
    if filters.get("type"):
        conditions += f" AND tte.type = '{filters.get('type')}'"
    
    if filters.get("branch"):
        conditions += f" AND tbu.branch = '{filters.get('branch')}'"
    
    if filters.get("currency"):
        conditions += f" AND tbu.currency = '{filters.get('currency')}'"
    
    if filters.get("member_status"):
        if filters.get("member_status") != "All":
            conditions += f" AND m.member_status = '{filters.get('member_status')}'"
    
    # Only get submitted documents
    conditions += " AND tbu.docstatus = 1"
    
    return conditions

def get_data(filters):
    """
    Fetch tithe transaction data from the parent doctype Tithe Batch Update.
    """
    conditions = get_conditions(filters)
    
    query = f"""
        SELECT
            tte.member_id,
            tte.full_name,
            m.member_status,
            tbu.branch,
            tbu.name as batch_name,
            tbu.reporting_date as date,
            tbu.currency,
            tte.type,
            tte.amount_paid,
            tte.exchange_rate,
            tte.worker_tithe,
            tte.member_tithe,
            tte.total_base_currency,
            tte.receipt_reference,
            tte.other_details
        FROM
            `tabTithe Batch Update` tbu
        INNER JOIN
            `tabTithe Transaction Entry` tte ON tte.parent = tbu.name
        LEFT JOIN
            `tabMember` m ON tte.member_id = m.name
        WHERE
            1=1
            {conditions}
        ORDER BY
            tte.member_id ASC,
            tbu.reporting_date DESC
    """
    
    data = frappe.db.sql(query, as_dict=1)
    
    if not data:
        return []
    
    return data

def add_subtotals_and_grand_totals(data, filters):
    """
    Add subtotal rows for each member and a grand total row at the end.
    """
    if not data:
        return []
    
    result = []
    current_member = None
    member_subtotal = {
        "total_base_currency": 0,
        "worker_tithe": 0,
        "member_tithe": 0,
        "amount_paid": 0,
        "transaction_count": 0
    }
    
    grand_total = {
        "total_base_currency": 0,
        "worker_tithe": 0,
        "member_tithe": 0,
        "amount_paid": 0,
        "transaction_count": 0,
        "member_count": 0
    }
    
    for row in data:
        if current_member != row.get("member_id"):
            if current_member is not None and member_subtotal["transaction_count"] > 0:
                result.append(get_subtotal_row(current_member, member_subtotal, data))
                grand_total["member_count"] += 1
            
            current_member = row.get("member_id")
            member_subtotal = {
                "total_base_currency": 0,
                "worker_tithe": 0,
                "member_tithe": 0,
                "amount_paid": 0,
                "transaction_count": 0
            }
        
        result.append(row)
        
        member_subtotal["total_base_currency"] += flt(row.get("total_base_currency", 0))
        member_subtotal["worker_tithe"] += flt(row.get("worker_tithe", 0))
        member_subtotal["member_tithe"] += flt(row.get("member_tithe", 0))
        member_subtotal["amount_paid"] += flt(row.get("amount_paid", 0))
        member_subtotal["transaction_count"] += 1
        
        grand_total["total_base_currency"] += flt(row.get("total_base_currency", 0))
        grand_total["worker_tithe"] += flt(row.get("worker_tithe", 0))
        grand_total["member_tithe"] += flt(row.get("member_tithe", 0))
        grand_total["amount_paid"] += flt(row.get("amount_paid", 0))
        grand_total["transaction_count"] += 1
    
    if current_member is not None and member_subtotal["transaction_count"] > 0:
        result.append(get_subtotal_row(current_member, member_subtotal, data))
        grand_total["member_count"] += 1
    
    if grand_total["transaction_count"] > 0:
        result.append(get_grand_total_row(grand_total))
    
    return result

def get_subtotal_row(member_id, subtotal, data):
    """
    Create a subtotal row for a member with styling indicators.
    """
    member_info = next((row for row in data if row.get("member_id") == member_id), {})
    
    return {
        "member_id": member_id,
        "full_name": member_info.get("full_name", ""),
        "member_status": member_info.get("member_status", ""),
        "batch_name": "",
        "branch": member_info.get("branch", ""),
        "type": "✨ SUBTOTAL ✨",
        "date": "",
        "currency": "",
        "amount_paid": subtotal["amount_paid"],
        "exchange_rate": "",
        "worker_tithe": subtotal["worker_tithe"],
        "member_tithe": subtotal["member_tithe"],
        "total_base_currency": subtotal["total_base_currency"],
        "receipt_reference": f"📊 {subtotal['transaction_count']} transactions",
        "other_details": "",
        "is_subtotal": 1,
        "indent": 1,
        "row_type": "subtotal"
    }

def get_grand_total_row(grand_total):
    """
    Create a grand total row with celebration styling.
    """
    return {
        "member_id": "🎯 GRAND TOTAL 🎯",
        "full_name": "",
        "member_status": "",
        "batch_name": "",
        "branch": "",
        "type": "🏆 GRAND TOTAL 🏆",
        "date": "",
        "currency": "",
        "amount_paid": grand_total["amount_paid"],
        "exchange_rate": "",
        "worker_tithe": grand_total["worker_tithe"],
        "member_tithe": grand_total["member_tithe"],
        "total_base_currency": grand_total["total_base_currency"],
        "receipt_reference": f"✨ {grand_total['member_count']} members • {grand_total['transaction_count']} transactions ✨",
        "other_details": "",
        "is_grand_total": 1,
        "indent": 0,
        "bold": 1,
        "row_type": "grand_total"
    }

def get_summary(data, filters):
    """
    Generate a cute summary for the report footer.
    """
    if not data:
        return []
    
    transaction_data = [row for row in data if not row.get("is_subtotal") and not row.get("is_grand_total")]
    
    if not transaction_data:
        return []
    
    total_base_currency = 0
    total_worker_tithe = 0
    total_member_tithe = 0
    total_amount_paid = 0
    unique_members = set()
    
    for row in transaction_data:
        total_base_currency += flt(row.get("total_base_currency", 0))
        total_worker_tithe += flt(row.get("worker_tithe", 0))
        total_member_tithe += flt(row.get("member_tithe", 0))
        total_amount_paid += flt(row.get("amount_paid", 0))
        unique_members.add(row.get("member_id"))
    
    summary = [
        {
            "value": len(unique_members),
            "label": _("👥 Members Contributed"),
            "datatype": "Int",
            "fieldname": "member_count",
            "currency": None
        },
        {
            "value": len(transaction_data),
            "label": _("📋 Total Transactions"),
            "datatype": "Int",
            "fieldname": "transaction_count",
            "currency": None
        },
        {
            "value": total_amount_paid,
            "label": _("💰 Total Amount Paid"),
            "datatype": "Currency",
            "fieldname": "total_amount_paid",
            "currency": filters.get("base_currency", "USD")
        },
        {
            "value": total_base_currency,
            "label": _("🏦 Total Tithe (Base)"),
            "datatype": "Currency",
            "fieldname": "total_base_currency",
            "currency": filters.get("base_currency", "USD")
        },
        {
            "value": total_worker_tithe,
            "label": _("🙏 Worker Tithe"),
            "datatype": "Currency",
            "fieldname": "total_worker_tithe",
            "currency": filters.get("base_currency", "USD")
        },
        {
            "value": total_member_tithe,
            "label": _("🏡 Member Tithe"),
            "datatype": "Currency",
            "fieldname": "total_member_tithe",
            "currency": filters.get("base_currency", "USD")
        }
    ]
    
    return summary